### ISSUE TITLE
High cpu usage and live reload not working

### ISSUE THREAD

**Issue by krepflap** [phoenixframework/phoenix#3365]
In dev env on Windows 10 with [WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10) the live reload doesn't work most of the time and I'm having high cpu usage after a while with beam.sp process maxing out remaining cpu usage.

So at first live reload works (first refresh when changing code), then beam.sp starts taking more of the CPU, eventually live reload and even webpack stop working (something to do with the notify events on the OS?). Other stuff works (rendering pages, ecto doing queries etc.), besides generally being slower because of the cpu usage.

I tracked it down by downgrading erlang versions.

The highest version where it doesn't happen is `es-erlang`  **1:21.1.4-1**.
Starting from the version `es-erlang` **1:21.2-1** (and higher) the issue happens. So some kind of regression between Erlang/OTP 21.1 and 21.2, but I don't know if it's Phoenix specific.

Currently I version locked `es-erlang` to 1:21.1.4-1 as a workaround (and that 100% solves the issue).

### Environment

* Elixir version (elixir -v): 1.8.1 
* Phoenix version (mix deps): 1.4.3
* NodeJS version (node -v): v8.15.1
* NPM version (npm -v): 6.4.1
* Operating system: Windows 10 16299.1087 with Ubuntu 18.04 WSL

### Expected behavior

Live reload to work with latest Erlang/OTP 21.2+ on WSL
Don't have beam.sp take up all remaining cpu usage

### Actual behavior

Live reload & webpack stop working after a while, they don't seem to notice fs notify events anymore and high cpu usage by beam.sp

**Comment by krepflap** [comment]
Confirmed that this doesn't happen on native Ubuntu 18.04.
Only in WSL in Windows 10.

**Comment by josevalim** [comment]
Hi @krepflap, webpack and live reload use different fs watching mechanisms, one in JS and another in Elixir, so if both are not working the issue is most likely on WSL. It may be that both are falling back to a polling mechanism, which would explain the high CPU usage and the high latency. According to this article everything is working fine though: https://devblogs.microsoft.com/commandline/wsl-adds-inotify-filesystem-change-notification-support/

Have you tried other backends or forcing a particular backend to see if the results changed? You can also investigate the file_system project and see if you find anything relevant: https://github.com/falood/file_system

**Comment by nickjj** [comment]
I run Windows 10 and WSL and run both Phoenix and Webpack through Docker (Docker for Windows).

This might not be 100% related but live reloading assets only works with Webpack if I turn on polling in my webpack config however CPU usage isn't spiking. I never got live reload to work with EEx templates (mainly because itnofiy events don't get passed through Docker containers on Windows, but that's a story for another time).

I guess I'm replying mainly to say that polling with Webpack doesn't seem to cause CPU spikes here.

**Comment by josevalim** [comment]
@krepflap if I understand correctly your comment on another issue, you were able to address this?

**Comment by krepflap** [comment]
@josevalim Only with a workaround: by version locking to Erlang/OTP 21.1 (apt-mark hold). If I upgrade to >= OTP 21.2, the problem happens again.

To reproduce, in WSL (Ubuntu 18.04 flavor) in Windows 10, with erlang/elixir/phoenix installed inside the Ubuntu (no docker):

```console
$ mix phx.new --no-ecto new
$ cd new
<fetch deps & build assets>
$ mix phx.server
```

Open browser http://localhost:4000, change something in app.css e.g. adding:

```css
h1 {
  color: red;
}
```

Browser autorefreshes and shows the expected change. When you continue to make changes (e.g. cycling colors, yellow, blue, green, etc.) the browser is supposed to refresh every time and show the change in color. This works with OTP 21.1.

With OTP 21.2 the live reload works once (first time after starting mix phx.server) then stops working at subsequent modifications. At which point I notice beam.smp capping out CPU.

Seems to be the same with vim (without plugins) or Visual Studio Code. I'll do some more checks.

**Comment by krepflap** [comment]
> [...] webpack and live reload use different fs watching mechanisms, one in JS and another in Elixir, so if both are not working the issue is most likely on WSL. 

Just to note that webpack works fine when run standalone e.g. `./node_modules/webpack/bin/webpack.js --mode development --watch`. It's picking up all the changes.

**Comment by krepflap** [comment]
@josevalim Ok I can reproduce it in isolation just with `file_system` (0.2.6), without Phoenix.

Please ignore my crappy code, I barely know what I'm doing. 🙇

```elixir
defmodule Fsdebug do
  @moduledoc """
  Documentation for Fsdebug.
  """

  use GenServer

  def start_link(args) do
    GenServer.start_link(__MODULE__, args)
  end

  def init(_args) do
    {:ok, watcher_pid} = FileSystem.start_link(dirs: ["/tmp"])
    FileSystem.subscribe(watcher_pid)
    {:ok, %{watcher_pid: watcher_pid}}
  end

  def handle_info({:file_event, watcher_pid, {path, events}}, %{watcher_pid: watcher_pid}=state) do
    IO.puts "Live reload: #{Path.relative_to_cwd(path)}"
    {:noreply, state}
  end
end
```

After launching that (inside WSL), I append a few times to a file under `/tmp`:
```console
$ for i in $(seq 1 20); do echo "koko" >> /tmp/koko; sleep 0.5; done
```

**Results:**

esl-erlang=1:21.1.4-1: works as expected, I see the output and everything is very responsive
esl-erlang=1:21.3.7-1: output stops halfway the iterations, beam capping out cpu

Erlang installed via repo http://binaries.erlang-solutions.com/debian

**Comment by josevalim** [comment]
Oh wow, great job. Can you please put that code in a simple repository and push it to github?

Also, do you know which FS backend you are using? And is `IO.puts "Live reload: #{Path.relative_to_cwd(path)}"` being invoked? How frequently?

**Comment by krepflap** [comment]
>Can you please put that code in a simple repository and push it to github?

Sure. https://github.com/krepflap/fsdebug

I just launch it like this (sorry I'm very new to Elixir):
```
$ iex -S mix
[...]
iex(1)> Fsdebug.start_link("")
```

Then in another terminal:
```
$ for i in $(seq 1 20); do echo "koko" >> /tmp/koko; sleep 0.5; done
```

>And is IO.puts [...] being invoked? How frequently?

Yes it's being invoked. When I talk about "output" in the results, that's what I mean (the IO.puts). With OTP 21.1 it outputs on every event as expected. With >= 21.2 it seems to output the first few events, then nothing anymore while CPU starts capping out.

I suspect it's a very specific issue though with the WSL kernel and that we won't be able to reproduce on anything else. Do you think it could have something to do with this change in Erlang OTP 21.2 ?

```

  OTP-15475    Application(s): erts
               Related Id(s): ERIERL-229

               Add a new pollset that is made to handle sockets that
               use {active, true} or {active, N}. The new pollset will
               not be polled by a pollthread, but instead polled by a
               normal scheduler.

               This change was made because of the overhead associated
               with constantly having to re-apply the ONESHOT
               mechanism on fds that all input events were
               interesting.

               The new pollset is only active on platforms that
               support concurrent kernel poll updates, i.e. Linux and
               BSD.
```

From http://erlang.org/download/otp_src_21.2.readme

>Also, do you know which FS backend you are using?

I'm pretty sure the backend should be inotify. Because the WSL does behave like Linux (the OS), it has inotify-tools installed etc. However the kernel isn't Linux.

From wikipedia:

>WSL provides a Linux-compatible kernel interface developed by Microsoft (containing no Linux kernel code),[1] which can then run a GNU user space on top of it, such as that of Ubuntu,[...]

**Comment by josevalim** [comment]
@krepflap thanks! Can you please help us isolate the issue a bit further? I am afraid it can indeed be a VM issue, so I want to make sure we have a minimal case to reproduce it. The next step would be to remove the `file_system` dependency. The difference is, instead of starting FileSYstem in your gen server, you can start the inotify port directly:

https://github.com/falood/file_system/blob/42d0a730c8df88a1aebe6d07d8360cdbf3c62461/lib/file_system/backends/fs_inotify.ex#L109-L116

And then your handle info can look like this:

```
  def handle_info({port, event}, %{port: port}=state) do
    IO.inspect event
    {:noreply, state}
  end
```

If you can reproduce it only with the port, then we can even translate it to Erlang and provide a minimal case. Thanks!

**Comment by krepflap** [comment]
Ok that works. I updated the code to remove the dependency : https://github.com/krepflap/fsdebug
I simplified to just have inotifywait watch modify events.

Results:

**esl-erlang (1:21.1.4-1):**

```
iex(1)> Fsdebug.start_link ""
{:ok, #PID<0.143.0>}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
iex(2)>
```

**esl-erlang (1:21.3.7-1):**

```
iex(1)> Fsdebug.start_link ""
{:ok, #PID<0.135.0>}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
{:data, {:eol, '/tmp/ MODIFY koko'}}
iex(2)>
```

Some missing events ^  (8 from 20 missing). No more output after that.

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 6637 xxxxx     20   0 2669248  32760   9604 S 400.0  0.2   0:18.39 beam.smp
```

High CPU usage ^

```
$ uname -a
Linux <snip> 4.4.0-43-Microsoft #1-Microsoft Wed Dec 31 14:42:53 PST 2014 x86_64 x86_64 x86_64 GNU/Linux
```

Almost seems like it's choking on multiple notify events after a while.

**Comment by josevalim** [comment]
@krepflap perfect. One last step before we can report this to the Erlang/OTP team. I have translated it to erlang:

```erlang
-module(fsdebug).
-compile(export_all).

start_link() ->
  gen_server:start_link(?MODULE, ok, []).

init(ok) ->
  BashArgs = ["-c", os:find_executable(inotify_wait) ++ " $0 $@ & PID=$!; read a; kill -KILL $PID"],
  PortArgs = ["-e", "modify", "--quiet", "-m", "-r", "/tmp"],
  Port = open_port(
    {spawn_executable, "/bin/sh"},
    [stream, exit_status, {line, 16384}, {args, BashArgs ++ PortArgs}, {cd, "/tmp"}]
  ),
  link(Port),
  process_flag(trap_exit, true),
  {ok, Port}.

handle_info({Port, Event}, Port) ->
  io:format("~p~n", [Event]),
  {noreply, Port}.
```

Please write that to a file named `fsdebug.erl`. And then do this:

```
$ erl
Erlang/OTP 21 [erts-10.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Eshell V10.0  (abort with ^G)
1> c("fsdebug").
fsdebug.erl:2: Warning: export_all flag enabled - all functions will be exported
{ok,fsdebug}
2> fsdebug:start_link().
```

Can you reproduce the issue once you start writing to the directory? If so, we have a minimal case that we can report to them.

**Comment by krepflap** [comment]
Yes still consistent with the erlang code. Had to make one modification to the BashArgs line though for it to work : 

```erlang
-module(fsdebug).
-compile(export_all).

start_link() ->
  gen_server:start_link(?MODULE, ok, []).

init(ok) ->
  BashArgs = ["-c", os:find_executable("inotifywait") ++ " $0 $@ & PID=$!; read a; kill -KILL $PID"],
  PortArgs = ["-e", "modify", "--quiet", "-m", "-r", "/tmp"],
  Port = open_port(
    {spawn_executable, "/bin/sh"},
    [stream, exit_status, {line, 16384}, {args, BashArgs ++ PortArgs}, {cd, "/tmp"}]
  ),
  link(Port),
  process_flag(trap_exit, true),
  {ok, Port}.

handle_info({Port, Event}, Port) ->
  io:format("~p~n", [Event]),
  {noreply, Port}.
```

With

```console
$ for i in $(seq 1 20); do echo "koko" >> /tmp/koko; sleep 0.5; done
```

Results (you can see `erts` version in output):

```
Erlang/OTP 21 [erts-10.1.3] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Eshell V10.1.3  (abort with ^G)
1> c("fsdebug.erl").
fsdebug.erl:2: Warning: export_all flag enabled - all functions will be exported
{ok,fsdebug}
2> fsdebug:start_link().
{ok,<0.85.0>}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
```

vs:

```
Erlang/OTP 21 [erts-10.3.4] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Eshell V10.3.4  (abort with ^G)
1> c("fsdebug.erl").
fsdebug.erl:2: Warning: export_all flag enabled - all functions will be exported
{ok,fsdebug}
2> fsdebug:start_link().
{ok,<0.86.0>}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
3>
```

No more output and high CPU.

**Comment by josevalim** [comment]
@krepflap one last question: if you replace the inotify_wait by something like `tail -f file` and start writing to that file something like echo "foo\n", does it reproduce the issue?

**Comment by josevalim** [comment]
Meanwhile, I created a bug report here: https://bugs.erlang.org/browse/ERL-933

Thanks for all the help so far! :heart:

**Comment by krepflap** [comment]
Thanks for bug report.

It's true that I can't reproduce this on native Linux e.g. Ubuntu 18.04 on AWS or my linux laptop. Only in WSL on Windows 10. In native Linux, the VM just purrs along happily with minimal cpu usage and stuff just works.

Just to note that it also happens on 21.2, so it's seems to be a regression going from 21.1 to 21.2.

```
Erlang/OTP 21 [erts-10.2.5] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Eshell V10.2.5  (abort with ^G)
1> c("fsdebug.erl").
fsdebug.erl:2: Warning: export_all flag enabled - all functions will be exported
{ok,fsdebug}
2> fsdebug:start_link().
{ok,<0.85.0>}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
{data,{eol,"/tmp/ MODIFY koko"}}
3>
```

(As you can see not all 20 events are being registered).

>if you replace the inotify_wait by something like tail -f file and start writing to that file something like echo "foo\n", does it reproduce the issue?

Yes. With this code and the same bash `for` loop, exactly the same behavior:

```erlang
-module(fsdebug).
-compile(export_all).

start_link() ->
  gen_server:start_link(?MODULE, ok, []).

init(ok) ->
  BashArgs = ["-c", os:find_executable("tail") ++ " $0 $@ & PID=$!; read a; kill -KILL $PID"],
  PortArgs = ["-f", "/tmp/koko"],
  Port = open_port(
    {spawn_executable, "/bin/sh"},
    [stream, exit_status, {line, 16384}, {args, BashArgs ++ PortArgs}, {cd, "/tmp"}]
  ),
  link(Port),
  process_flag(trap_exit, true),
  {ok, Port}.

handle_info({Port, Event}, Port) ->
  io:format("~p~n", [Event]),
  {noreply, Port}.
```

>Thanks for all the help so far!

No problem, great way to learn stuff.

**Comment by josevalim** [comment]
Perfect, thanks. Since this is not a Phoenix issue, I will go ahead and close this, we can continue the discussion in Erlang's issues tracker.
